### PR TITLE
Reapply "[AMDGPU] Always lower s/udiv64 by constant to MUL"

### PR DESCRIPTION
**Pull request by Pierre-vh**
Reland #100723, fixing the ARM issue at the cost of a small regression in AMDGPU.

Solves #100383

### CODE REVIEW

**Comment by Pierre-vh** [comment]
@MaskRay please check that this is good for Halide now. The test case is based on the reproducer given.
@davemgreen please check the ARM test case

**Review by arsenm — APPROVED** [review]
Where is the AMDGPU regression? can it be fixed?

**Review by MaskRay — APPROVED** [review]
LGTM.
The internal Halide ARM target still passes with this patch.

@davemgreen might want to check the naming of the test llvm/test/CodeGen/ARM/div-by-constant-to-mul-crash.ll
I think it's too specific to use a "*-crash" to name the test.
One of the existing `*div*.ll` tests might be better.

File: llvm/test/CodeGen/ARM/div-by-constant-to-mul-crash.ll [hunk 1]
```diff
@@ -0,0 +1,56 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=arm--linux-gnueabihf -mcpu= -mattr=+neon | FileCheck %s
```
Comment (davemgreen) [inline llvm/test/CodeGen/ARM/div-by-constant-to-mul-crash.ll:2]: I'm not sure what -mcpu= does, but can it be removed? And can you use armv7--linux-gnueabihf if it still shows the same issue.

**Review by davemgreen — COMMENTED** [review]
The ARM test LGTM which a minor suggestion to update the triple used. Thanks.

**Comment by Pierre-vh** [comment]
> Where is the AMDGPU regression? can it be fixed?

Some small changes in `fshr.ll` (current version is on the left)
image

File: llvm/test/CodeGen/ARM/div-by-constant-to-mul-crash.ll (hunk 1 above)
Comment (Pierre-vh) [reply llvm/test/CodeGen/ARM/div-by-constant-to-mul-crash.ll:2]: done

**Pull request merged**